keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,682 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a generic feature to protect objects from deletion while it is marked as in-use.
+
+## Motivation
+
+Currently, "a generic mechanism that prevents a resource from being deleted when it is in-use" doesn't exist.
+Instead, for such a use case, each controller, like [pv-protection](https://github.com/kubernetes/enhancements/issues/499) and [pvc-protection](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/postpone-pvc-deletion-if-used-in-a-pod.md), implements its own logic to protect used objects from being deleted while in-use.
+These controllers use [Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/) to block deletion.
+Finalizers block the _completion_ of a delete operation, but they do not prevent the deletion from starting.
+Once started, a delete _will_ complete and can not be aborted.
+
+However, Finalizer may not be appropriate for some resource protection use cases, because it doesn't prevent other controllers from executing their pre-deletion actions.
+The order of execution of pre-deletion actions across finalizers is not defined.
+As a result, other controllers will execute their pre-deletion actions while a finalizer for protection still exists.
+
+This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
+- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+- Implement any of the above potential use cases
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
```

Comment:
I hope we can find a more human friendly name :)

UserFinalizer or ActiveUsage or or Utilizers

Reply:
Actually the name came from [here](https://github.com/kubernetes/kubernetes/issues/10179#issuecomment-580932333), but `Utilizers` looks nice to me, as a non-native speaker of English. I will wait for feedback from others (I think that sig-api-machinery has been called this concept as Liens for a long time, so they may have their reasons for the name).

Reply:
"Lien" is a really tight analogy to what this is doing, which is probably why @bgrant0607 chose the word initially. Everyone with a mortgage should understand what a lien is; that's a lot of people, though certainly not everyone.

I personally hate the word "utilize" because we already have the perfectly good "use", but unfortunately "users" would not be very clear.

If we can't use "lien" the next word on my list would probably be "Reservations" or "Reservers".

Reply:
> Everyone with a mortgage should understand what a lien is; that's a lot of people, though certainly not everyone.

It is exactly what "as a non-native speaker of English" meant. Actually, I used dictionary when I first heard the name of the concept. Although, it might be just due to the lack of my vocabulary.

Reply:
I didn't have any trouble understanding lien in this context.  Users is likely to be confused with "people using kube".  Utilizers seems inaccurate because this is not necessarily tracking the actual resource which has declared, "don't delete this"

Reply:
How about "InUse"?

Reply:
I like the word "liens" for this - it's very specific and maps very closely to the meaning.

if we NEED to go chattier, the next most common phrase I hear is "deleteInhibitors".  That's a lot more letters for no additional connotation.

It's rare that we find a "just right" word.

Reply:
The potential for confusion around `lein` in casual usage (homonyms), it's relatively niche usage in a business context (I find the mortgage argument unconvincing simply because I bet most homeowners *don't* think of the word "liens"), and potentially negative connotations around the word (although I rate this the least significant).

I was hoping that some of the justification for the name would come along with the KEP, vs being left as a blank space for me to guess at what else has been tried :)

The most succinct human definition I've seen so far is:

> A consumer of this resource considers its removal to have a significant negative impact on correct operation, and furthermore desires that any other human or automation interacting with the resource NOT delete it unless said client is fully aware of the implications (which roughly revolve around either major disruption or an unrecoverable state)

Is that a fair summation?

It doesn't necessarily imply the resource is in use, and I agree that "users" is also potentially confusing from the perspective of humans vs machines.  It does argue that a particular client wants to add an additional level of importance to a resource to distinguish between "replaceable automatically without major impact" (typically implied by the controller field) vs "not replaceable automatically or without major impact".

Some other terms that potentially might make sense (would like to see arguments for/against for each):

* `hold`
* `lease`
* `claim`

Lein is typically defined as "hold on property", and I agree that this is very much a hold on something that is the property (owned by) a human or machine.  It is not "controlled", and I don't think the word specifically here is "ownership".

Reply:
`deletionBlockers`. Everybody will understand that.

As a non-native speaker I have never heard "liens".

Reply:
As non-native speaker, I've never heard "liens" either.
But I must admit that after translating it back then when I first seen that, I actually like this

Reply:
As another non-native I'm in favor of `liens` as well, `deletionBlockers` is pretty mouthful and the former translates pretty well.

Reply:
I'm thinking of a number between 0 and 3^32.  Whoever guesses closest gets to choose the name from the following list:

1) liens
2) deletionBlockers
3) holds
4) deleteInhibitors

Go.

Reply:
I'm kidding of course.  I still like "liens" but I can live with "holds".

Reply:
Chipping in from SIG Docs. To help us decide, would someone like to draft an entry in the Kubernetes glossary that includes an initial definition of “lien”?

Reply:
Lien: An indication that some entity is relying on an API object.  Deletion of objects with liens will be prevented until the interested party releases the lien.